src/Compilers/CSharp/Portable/Emitter/Model/PEAssemblyBuilder.cs

```diff
@@ -131,6 +150,66 @@ protected override void AddEmbeddedResourcesFromAddedModules(ArrayBuilder<Cci.Ma
+            {
+                return new SynthesizedAttributeData(constructor, ImmutableArray<TypedConstant>.Empty, ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+            }
+
+            return base.SynthesizeEmbeddedAttribute();
+        }
+
+        internal override SynthesizedAttributeData SynthesizeIsReadOnlyAttribute()
+        {
+            if (_embeddedAttributesConstructors.TryGetValue(WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor, out var constructor))
+            {
+                return new SynthesizedAttributeData(constructor, ImmutableArray<TypedConstant>.Empty, ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+            }
+
+            return base.SynthesizeIsReadOnlyAttribute();
+        }
+
+        private ImmutableDictionary<WellKnownMember, MethodSymbol> CreateEmbeddedAttributesIfNeeded(DiagnosticBag diagnostics)
+        {
+            var builder = ArrayBuilder<KeyValuePair<WellKnownMember, MethodSymbol>>.GetInstance();
+
+            if (_sourceAssembly.DeclaringCompilation.NeedsGeneratedIsReadOnlyAttribute)
+            {
+                CreateEmbeddedAttributeIfNeeded(
+                    builder,
+                    diagnostics,
+                    WellKnownType.Microsoft_CodeAnalysis_EmbeddedAttribute,
+                    WellKnownMember.Microsoft_CodeAnalysis_EmbeddedAttribute__ctor,
+                    allowUserDefined: false);
+
+                CreateEmbeddedAttributeIfNeeded(
+                    builder,
+                    diagnostics,
+                    WellKnownType.System_Runtime_CompilerServices_IsReadOnlyAttribute,
+                    WellKnownMember.System_Runtime_CompilerServices_IsReadOnlyAttribute__ctor,
+                    allowUserDefined: true);
+            }
+
+            return ImmutableDictionary.CreateRange(builder.ToArrayAndFree());
+        }
+
+        private void CreateEmbeddedAttributeIfNeeded(ArrayBuilder<KeyValuePair<WellKnownMember, MethodSymbol>> builder, DiagnosticBag diagnostics, WellKnownType type, WellKnownMember constructor, bool allowUserDefined)
+        {
+            var userDefinedAttribute = _sourceAssembly.DeclaringCompilation.GetWellKnownType(type);
+            if (userDefinedAttribute == null ||
+                userDefinedAttribute is MissingMetadataTypeSymbol ||
+                !userDefinedAttribute.ContainingAssembly.Equals(_sourceAssembly))
+            {
+                var attributeSymbol = new SourceEmbeddedAttributeSymbol(type, _sourceAssembly.DeclaringCompilation);
+                builder.Add(new KeyValuePair<WellKnownMember, MethodSymbol>(constructor, attributeSymbol.Constructor));
+            }
+            else if (!allowUserDefined)
+            {
+                diagnostics.Add(ErrorCode.ERR_TypeReserved, userDefinedAttribute.Locations[0], userDefinedAttribute);
```

Comment:
I think that for the purpose of this error we should use location of the conflicting type only if it is declared in the source module. Conflict with types from added modules and NoPia embedded types should be handled in a different place, I think we can just rely on existing mechanism for that.

---
In reply to: [113573365](https://github.com/dotnet/roslyn/pull/18715#discussion_r113573365) [](ancestors = 113573365)